gateway/transport/plugins/audit/wal.go

```diff
@@ -81,22 +81,66 @@ func (p *auditPlugin) dropWalLog(sid string) {
 	walogm.mu.Unlock()
 }
 
-func (p *auditPlugin) writeOnClose(pctx plugintypes.Context, errMsg error) error {
-	walLogObj := p.walSessionStore.Pop(pctx.SID)
-	walogm, ok := walLogObj.(*walLogRWMutex)
-	if !ok {
-		return nil
+func (p *auditPlugin) flushLogsFromDisk(pctx plugintypes.Context, errMsg error) error {
+	walFolder := fmt.Sprintf(walFolderTmpl, plugintypes.AuditPath, pctx.OrgID, pctx.SID)
+	// Open the WAL log from disk and get the header
+	walog, wh, err := sessionwal.OpenWithHeader(walFolder)
+
+	walogm := &walLogRWMutex{
+		log:        walog,
+		mu:         sync.RWMutex{},
+		folderName: walFolder,
 	}
+
 	walogm.mu.Lock()
 	defer func() { _ = walogm.log.Close(); walogm.mu.Unlock() }()
-	// we could add an attribute to have the last message
-	// propagated as metadata instead inside the stream
-	if errMsg != nil && errMsg != io.EOF {
-		err := walogm.log.Write(eventlogv1.New(time.Now().UTC(), eventlogv1.ErrorType, []byte(errMsg.Error()), nil))
+	if err != nil {
```

Comment:
You must validate the error before trying to do anything with walog or wh instance